### ISSUE TITLE
Allow custom configs `params` to be parsed before nextflow.config

### ISSUE THREAD

**Issue by mahesh-panchal** [nextflow-io/nextflow#2662]
## New feature

I'm not sure whether to classify this as a feature request or bug.
The issue is that `params` configuration from custom configs (`-c`) appears to be read in after nextflow.config. 
The docs describe that `params` provided in a custom config have a higher priority than those in the nextflow.config. This is true at the time of running the workflow, but not when parsing and evaluating the conditions in the Nextflow config.

For example:
`main.nf`:
```nextflow
#! /usr/bin/env nextflow

nextflow.enable.dsl = 2

workflow {
    FOO()
}

process FOO {

    output:
    path 'myfile.txt'

    script:
    """
    touch myfile.txt
    """
}
``` 
`nextflow.config`:
```nextflow
params.outdir = 'results'
process {
    withName: 'FOO' {
        publishDir = "$params.outdir/foo"
    }
}
```

If one runs:
```console
$ nextflow run main.nf --outdir '/my/new/path'
```
or 
```console
$ cat params.yml
outdir : '/my/new/path'
$ nextflow run main.nf -params-file params.yml
```
Then the `publishDir` is correctly set.

However, if one uses
```console
$ cat custom.config
params.outdir = '/my/new/path'
$ nextflow run main.nf -c custom.config
```
Then the `publishDir` is not correct, instead using the default value.

## Usage scenario 

Lot's of Nextflow users currently provide custom `params` via the `-c` option, not realizing it is not applying to configuration in the nextflow.config.

## Suggest implementation 

(Highlight the main building blocks of a possible implementation and/or related components)

**Comment by abhi18av** [comment]
Good point, this situation seems to also affect launches from the `tw` CLI 


```powershell

tw launch $PIPELINE_URL `
    --profile conda_local `
    --params-file params.yml `              <-- Local params file
    --config slurm.config `                 <-- Local config file
    --compute-env $COMPUTE_ENV_ID `

```

**Comment by mahesh-panchal** [comment]
Is this intended to be addressed in the new config implementation?

**Comment by abhi18av** [comment]
@mahesh-panchal , I think that this would also be addressed as per the overall vision. Though I don't have any more visibility at this point

**Comment by maxulysse** [comment]
I managed to overcome this by supplying params in a yml file

**Comment by mahesh-panchal** [comment]
Not everyone knows to do this. Also configs support things the yml and command line do not, e.g. environment variables (not possible in yml), or closures (config specific).

**Comment by maxulysse** [comment]
> Not everyone knows to do this. Also configs support things the yml and command line do not, e.g. environment variables (not possible in yml), or closures (config specific).

I know, I was commenting more to make the issue not stale anymore

**Comment by jfy133** [comment]
Could someone please document this mysterious `params` yml file :sweat

**Comment by jfy133** [comment]
Oh btw it should be noted this issue only applies for DSL2 workflows

**Comment by jfy133** [comment]
The documentation has been partly addressed (, see the closed PR above which was replaced by a different PR by the main Devs), but I think this is still an issue

**Comment by bentsherman** [comment]
I realized I've been seeing this issue for a while with nf-core pipelines. As described in the OP, if you set `params.outdir` in a config file instead of on the command line or params file, Nextflow will save some reports to a `null` directory while everything else will be published correctly.

In order to fix it, I think we would need to parse the config files twice -- first to collect the params, second to parse the other scopes. I'm not sure how difficult that would be.

~~@pditommaso I think this issue is another good reason to allow the report file settings to be closures (#3651). This way, the settings would be evaluated only after all the params and config files have been resolved.~~

**Comment by mahesh-panchal** [comment]
What's the current thinking on this?

**Comment by bentsherman** [comment]
For now I'm trying to work around this issue instead of solving it directly, because I'm not sure if I want to go down the path of lazily resolving all params used in the config file.

For example, the case of publishDir is addressed by workflow outputs, since the output directory is specified once in the main config:
```groovy
// override with -output-dir CLI option
outputDir = 'results'
```

There is a similar issue with report file names that we discussed in #5543

I think params should be overridden via command line or params file as a best practice, rather than through an extra config. I'd like to see if there are any other use cases where the extra config is necessary

**Comment by mahesh-panchal** [comment]
One thought is perhaps to take params out of the config. Supply default values through a special yaml/json file. This might make it easier for pipeline users to form inputs as there's already a file with the intended structure that they can then feed to -params-file. Validation is perhaps made easier too then.

The disadvantage then is not being able to use closures, etc in params. However then we really need a mechanism to allow users to define resource scaling functions that's accessible from the config. Under the strict syntax, that's only accessible from params. 

Just a thought.
